src/txmempool.cpp

```diff
@@ -99,96 +139,338 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
             // If recursively removing but origTx isn't in the mempool
             // be sure to remove any children that are in the pool. This can
-            // happen during chain re-orgs if origTx isn't re-accepted into
+            // happen during chain re-orgs if origCert isn't re-accepted into
             // the mempool for any reason.
-            for (unsigned int i = 0; i < origTx.vout.size(); i++) {
+            for (unsigned int i = 0; i < origTx.GetVout().size(); i++) {
                 std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(COutPoint(origTx.GetHash(), i));
                 if (it == mapNextTx.end())
                     continue;
-                txToRemove.push_back(it->second.ptx->GetHash());
+                LogPrint("sc", "%s():%d - adding tx [%s] to list for removing\n", __func__, __LINE__, it->second.ptx->GetHash().ToString());
+                objToRemove.push_back(it->second.ptx->GetHash());
+            }
+
+            for(const auto& sc: origTx.GetVscCcOut()) {
+                if (mapSidechains.count(sc.scId) == 0)
+                    continue;
+                if (removeDependantFwds) {
+                    for(const auto& fwdTxHash : mapSidechains.at(sc.scId).fwdTransfersSet)
+                        objToRemove.push_back(fwdTxHash);
+                } else
+                    objToRemove.push_back(mapSidechains.at(sc.scId).scCreationTxHash);
             }
         }
-        while (!txToRemove.empty())
+        removeInternal(objToRemove, removedTxs, removedCerts, fRecursive, removeDependantFwds);
+    }
+}
+
+void::CTxMemPool::removeInternal(
+    std::deque<uint256>& objToRemove,
+    std::list<CTransaction>& removedTxs,
+    std::list<CScCertificate>& removedCerts,
+    bool fRecursive,
+    bool removeDependantFwds)
+{
+    // called with lock taken
+    AssertLockHeld(cs);
+
+    while (!objToRemove.empty())
+    {
+        uint256 hash = objToRemove.front();
+        objToRemove.pop_front();
+        if (mapTx.count(hash))
         {
-            uint256 hash = txToRemove.front();
-            txToRemove.pop_front();
-            if (!mapTx.count(hash))
-                continue;
             const CTransaction& tx = mapTx[hash].GetTx();
             if (fRecursive) {
-                for (unsigned int i = 0; i < tx.vout.size(); i++) {
+                for (unsigned int i = 0; i < tx.GetVout().size(); i++) {
                     std::map<COutPoint, CInPoint>::iterator it = mapNextTx.find(COutPoint(hash, i));
                     if (it == mapNextTx.end())
                         continue;
-                    txToRemove.push_back(it->second.ptx->GetHash());
+                    objToRemove.push_back(it->second.ptx->GetHash());
+                }
+                for(const auto& sc: tx.GetVscCcOut()) {
+                    if (mapSidechains.count(sc.scId) == 0)
+                        continue;
+
+                    if (removeDependantFwds) {
```

Comment:
instead of if else just directly schedule removal of all txs that where sending funds to this sc
 for(const auto& fwdTxHash : mapSidechains.at(sc.scId).fwdTransfersSet)
                        objToRemove.push_back(fwdTxHash);

Reply:
Done